crates/bevy_ecs/src/bundle.rs

```diff
@@ -0,0 +1,233 @@
+}
+
+all_tuples!(tuple_impl, 0, 15, C);
+
+#[derive(Debug, Clone, Copy)]
+pub struct BundleId(usize);
+
+impl BundleId {
+    #[inline]
+    pub fn index(self) -> usize {
+        self.0
+    }
+}
+
+impl SparseSetIndex for BundleId {
+    #[inline]
+    fn sparse_set_index(&self) -> usize {
+        self.index()
+    }
+
+    fn get_sparse_set_index(value: usize) -> Self {
+        Self(value)
+    }
+}
+
+pub struct BundleInfo {
+    pub(crate) id: BundleId,
+    pub(crate) component_ids: Vec<ComponentId>,
+    pub(crate) storage_types: Vec<StorageType>,
+}
+
+impl BundleInfo {
+    /// # Safety
+    /// table row must exist, entity must be valid
+    #[inline]
+    pub(crate) unsafe fn put_components<T: DynamicBundle>(
+        &self,
+        sparse_sets: &mut SparseSets,
+        entity: Entity,
+        table: &Table,
+        table_row: usize,
+        bundle_flags: &[ComponentFlags],
+        bundle: T,
+    ) {
+        // NOTE: put is called on each component in "bundle order". bundle_info.component_ids are also in "bundle order"
+        let mut bundle_component = 0;
+        bundle.put(|component_ptr| {
+            // SAFE: component_id was initialized by get_dynamic_bundle_info
+            let component_id = *self.component_ids.get_unchecked(bundle_component);
+            let flags = *bundle_flags.get_unchecked(bundle_component);
+            match self.storage_types[bundle_component] {
+                StorageType::Table => {
+                    let column = table.get_column(component_id).unwrap();
+                    column.set_unchecked(table_row, component_ptr);
+                    column.get_flags_unchecked_mut(table_row).insert(flags);
+                }
+                StorageType::SparseSet => {
+                    let sparse_set = sparse_sets.get_mut(component_id).unwrap();
+                    sparse_set.insert(entity, component_ptr, flags);
+                }
+            }
+            bundle_component += 1;
+        });
+    }
+
+    #[inline]
+    pub fn id(&self) -> BundleId {
+        self.id
+    }
+
+    #[inline]
+    pub fn components(&self) -> &[ComponentId] {
+        &self.component_ids
+    }
+
+    #[inline]
+    pub fn storage_types(&self) -> &[StorageType] {
+        &self.storage_types
+    }
+}
+
+#[derive(Default)]
+pub struct Bundles {
+    bundle_infos: Vec<BundleInfo>,
+    bundle_ids: HashMap<TypeId, BundleId>,
+}
+
+impl Bundles {
+    #[inline]
+    pub fn get(&self, bundle_id: BundleId) -> Option<&BundleInfo> {
+        self.bundle_infos.get(bundle_id.index())
+    }
+
+    #[inline]
+    pub fn get_id(&self, type_id: TypeId) -> Option<BundleId> {
+        self.bundle_ids.get(&type_id).cloned()
+    }
+
+    pub(crate) fn init_info_dynamic<T: DynamicBundle>(
+        &mut self,
+        components: &mut Components,
+        bundle: &T,
+    ) -> &BundleInfo {
+        let bundle_infos = &mut self.bundle_infos;
+        let id = self.bundle_ids.entry(TypeId::of::<T>()).or_insert_with(|| {
```

Comment:
All Bundles are DynamicBundles, but DynamicBundles are not Bundles. `EntityMut::insert_bundle` takes a `DynamicBundle` and needs to "initialize the bundle info". It cannot call `init_bundle::<T: Bundle>` because `T` is a `DynamicBundle`,.

Reply:
I don't get why this split exists right now. All types implement both traits.

Reply:
Interesting callout. Hecs used the split, which got grandfathered in. My personal rationale was "a combined Bundle cannot be boxed due to static methods". However we don't have any need for boxed Bundles. Removing the split compiles just fine.